feat(curriculum): Add fCC Authors Page Project to JS Cert (#49826)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-fetch-and-promises-by-building-an-fcc-authors-page/641daabed8d0584b1150c953.md

---
id: 641daabed8d0584b1150c953
title: Step 28
challengeType: 0
dashedName: step-28
---

# --description--

> Now it's time to extract the part of the `bio` for uniformity. If the `bio` text is more than 50 characters, then you should extract the first 50 and replace the rest with an ellipsis. The `slice` method is ideal for this use case. Apart from using it to extract the part of an array you want, you can also use it on strings to do the same thing.
> 
> You might be thinking you can use `if` statement here, but you can't put an `if` statement inside a template literal. What you have to use instead is a ternary operator, AKA one line `if...else` statement.
> 
> Extract the first 50 letters of the authors' `bio` text and replace the rest with ellipsis (...). For the else statement, use the `bio` itself.

Reviewer: ```suggestion
Some of the author bios are much longer than others. To give the cards a uniform look, you can extract the first 50 characters of each one and replace the rest with an ellipsis (...). Otherwise, you can show the entire bio.

Within the paragraph element, replace `bio` with a ternary operator. For the condition, check if the length of `bio` is greater than 50. If it is, use the `.slice()` method to extract the first 50 characters of `bio` and add an ellipsis at the end. Otherwise, show the full `bio`.
```

  height: 15px;
}

.author-name {
  margin: 10px;
}

.bio {
  margin: 20px;
}

.error-msg {
  color: var(--light-grey);
}

.btn {
  cursor: pointer;
  width: 200px;
  margin: 10px;
  color: var(--main-bg-color);
  font-size: 14px;
  background-color: var(--golden-yellow);
  background-image: linear-gradient(#fecc4c, #ffac33);
  border-color: var(--golden-yellow);
  border-width: 3px;
}
```

```js
const authorContainer = document.getElementById('author-container');
const loadMoreBtn = document.getElementById('load-more-btn');

let startingIndex = 0;
let endingIndex = 8;
let authorDataArr = [];

fetch('https://cdn.freecodecamp.org/curriculum/news-author-page/authors.json')
  .then((res) => res.json())
  .then((data) => {
    authorDataArr = data;
    displayAuthors(authorDataArr.slice(startingIndex, endingIndex));  
  })
  .catch((err) => {
    console.error(`There was an error: ${err}`);
  });

const fetchMoreAuthors = () => {
  startingIndex += 8;
  endingIndex += 8;

  displayAuthors(authorDataArr.slice(startingIndex, endingIndex));
  if (authorDataArr.length <= endingIndex) {
    loadMoreBtn.disabled = true;
    loadMoreBtn.textContent = 'No more data to load';
  }
};

> const displayAuthors = (authors) => {
>   authors.forEach( ({ author, image, url, bio }, index) =>
>   (authorContainer.innerHTML += `
>     <div id="${index}" class="user-card">
>       <h2 class="author-name">${author}</h2>
>       <img class="user-img" src="${image}" alt="${author} avatar">
>       <div class="purple-divider"></div>
>       --fcc-editable-region--
>       <p class="bio">${bio}</p>
>       --fcc-editable-region--
>       <a class="author-link" href="${url}" target="_blank">${author} author page</a>
>     </div>
>   `)
>   )
> };

Reviewer: ```suggestion
const displayAuthors = (authors) => {
  authors.forEach(({ author, image, url, bio }, index) => {
    authorContainer.innerHTML += `
    <div id="${index}" class="user-card">
      <h2 class="author-name">${author}</h2>
      <img class="user-img" src="${image}" alt="${author} avatar">
      <div class="purple-divider"></div>
      --fcc-editable-region--
      <p class="bio">${bio}</p>
      --fcc-editable-region--
      <a class="author-link" href="${url}" target="_blank">${author}'s author page</a>
    </div>
  `;
  });
};
```

---
id: 641daabed8d0584b1150c953
title: Step 28
challengeType: 0
dashedName: step-28
---

# --description--

Some of the author bios are much longer than others. To give the cards a uniform look, you can extract the first 50 characters of each one and replace the rest with an ellipsis (...). Otherwise, you can show the entire bio.

Within the paragraph element, replace `bio` with a ternary operator. For the condition, check if the length of `bio` is greater than 50. If it is, use the `.slice()` method to extract the first 50 characters of `bio` and add an ellipsis at the end. Otherwise, show the full `bio`.

# --hints--

You should create a `p` element.

```js
assert.exists(document.querySelector('p'));
```

Your `p` element should have the class `bio`

```js
assert.match(code, /<p\s*class=("|')bio\1>/);
```

> You should check if the length of the `bio` text is greater than `50`. 
> 
> ```js
> assert.match(code, /<p\s*class=("|')bio\1>\$\{bio\.length\s+>\s+50/)
> ```
> 
> If the `bio` text is greater than `50` characters, you should extract the first 50 characters with `slice()` and replace the rest with `...`. Don't forget that indexes are zero-based.
> 
> ```js
> assert.match(code, /<p\s*class=("|')bio\1>\$\{bio\.length\s+>\s+50\s+\?\s+bio\.slice\(0\,\s+49\)\s+\+\s+("|')\.\.\.\2\s+\:/)
> ```
> 
> If the `bio` text is less than 50 characters, use the `bio` text directly.
> 
> ```js
> assert.match(code, /<p\s*class=("|')bio\1>\$\{bio\.length\s+>\s+50\s+\?\s+bio\.slice\(0\,\s+49\)\s+\+\s+("|')\.\.\.\2\s+\:\s+bio\}<\/p>/)
> ```

Reviewer 2: ````suggestion
You should check if the length of the `bio` text is greater than `50`. 

```js
assert.match(code, /<p\s*class=("|')bio\1>\$\{\s*bio\.length\s*>\s*50/)
```

If the `bio` text is greater than `50` characters, you should extract the first 50 characters with `slice()` and replace the rest with `...`. Don't forget that indexes are zero-based.

```js
assert.match(code, /<p\s*class=("|')bio\1>\$\{\s*bio\.length\s*>\s*50\s*\?\s*bio\.slice\(\s*0\,\s*49\s*\)\s*\+\s*("|')\.\.\.\2\s*\:/)
```

If the `bio` text is less than 50 characters, use the `bio` text directly.

```js
assert.match(code, /<p\s*class=("|')bio\1>\$\{\s*bio\.length\s*>\s*50\s*\?\s*bio\.slice\(\s*0\,\s*49\s*\)\s*\+\s*("|')\.\.\.\2\s*\:\s*bio\s*\}<\/p>/)
```
````

All of the spaces are optional here.
